windows/code/DBStorage.cpp

```diff
@@ -351,174 +339,236 @@ winrt::Windows::Foundation::IAsyncAction DBStorage::RunTasks()
                 co_return;
             }
             std::swap(tasks, m_tasks);
-            db = m_db;
+            db = m_db.get();
         }
 
         for (auto &task : tasks) {
-            task.Run(db);
-            if (cancellationToken())
-                break;
+            if (!cancellationToken()) {
+                task->Run(*this, db);
+            } else {
+                task->Cancel();
+            }
         }
     }
-    winrt::slim_lock_guard guard(m_lock);
-    m_action = nullptr;
-    m_cv.notify_all();
 }
 
-void DBStorage::DBTask::Run(sqlite3 *db)
+// Add new Error to the error list.
+// Return std::nullopt for convenience to other methods that use std::nullopt to indicate error
+// result.
+std::nullopt_t DBStorage::ErrorHandler::AddError(std::string &&message) noexcept
 {
-    switch (m_type) {
-        case Type::multiGet:
-            multiGet(db);
-            break;
-        case Type::multiSet:
-            multiSet(db);
-            break;
-        case Type::multiRemove:
-            multiRemove(db);
-            break;
-        case Type::clear:
-            clear(db);
-            break;
-        case Type::getAllKeys:
-            getAllKeys(db);
-            break;
-    }
+    m_errors.push_back(Error{std::move(message)});
+    return std::nullopt;
 }
 
-void DBStorage::DBTask::multiGet(sqlite3 *db)
+const std::vector<DBStorage::Error> &DBStorage::ErrorHandler::GetErrors() const noexcept
 {
-    if (!CheckArgs(db, m_args, m_callback)) {
-        return;
+    return m_errors;
+}
+
+DBStorage::DBTask::DBTask(DBStorage::ErrorHandler &errorHandler,
+                          std::function<void(DBTask &task, sqlite3 *db)> &&onRun) noexcept
+    : m_errorHandler(errorHandler), m_onRun(std::move(onRun))
+{
+}
+
+void DBStorage::DBTask::Run(DBStorage &storage, sqlite3 *db) noexcept
+{
+    if (!db) {
+        // We initialize DB handler on demand to report errors in the task context.
+        if (auto res = storage.InitializeStorage(m_errorHandler)) {
+            db = *res;
+        }
+    }
+    if (db) {
+        m_onRun(*this, db);
     }
```

Comment:
Should we log an error message if there is no db here or will that be too spammy?

Reply:
I am not sure if we log other error message. Let me check.

Reply:
I looked at the code. If there is no db here, then it means that we failed to initialize it earlier in the `InitializeStorage` method above. This method adds any related errors to the ErrorManager (replaced ErrorHandler in latest code.)